File: frontend/src/types/project.ts
export type CelebrityMode = "fictional_only" | "celebrity_inspired_only" | "licensed_real_celebrity_only";

export interface ProjectCreateInput {
  target_original_video_url: string;
  example_original_video_url: string;
  example_remix_video_url: string;
  character_style: string;
  region_style_swap: string;
  gender_mix: string;
  age_group: string;
  ethnic_cultural_direction: string;
  celebrity_mode: CelebrityMode;
  visual_theme: string;
  costume_style: string;
  lighting_style: string;
  cinematic_mood: string;
  dance_style: string;
  energy_level: string;
  camera_style: string;
  preserve_melody: boolean;
  remix_genre: string;
  beat_intensity: string;
  vocal_handling: string;
}

export interface CharacterBible {
  cast_name: string;
  aliases: string[];
  persona_summary: string;
  styling_notes: string[];
  movement_notes: string[];
}

export interface Scene {
  scene_number: number;
  title: string;
  setting: string;
  visual_focus: string;
  choreography_note: string;
}

export interface ScenePrompt {
  scene_number: number;
  prompt: string;
}

export interface ProjectSummary {
  id: number;
  target_original_video_url: string;
  remix_genre: string;
  celebrity_mode: CelebrityMode;
  status: string;
  created_at: string;
}

export interface ProjectDetail extends ProjectSummary, ProjectCreateInput {
  config_json: Record<string, unknown> | null;
  transformation_summary: string | null;
  character_bible: CharacterBible | null;
  storyboard_scenes: Scene[] | null;
  scene_prompts: ScenePrompt[] | null;
  editing_plan: string[] | null;
  consistency_rules: string[] | null;
  manifest: Record<string, unknown> | null;
  updated_at: string;
}

export interface ProjectPlan {
  transformation_summary: string;
  character_bible: CharacterBible;
  storyboard_scenes: Scene[];
  scene_prompts: ScenePrompt[];
  editing_plan: string[];
  consistency_rules: string[];
  manifest: Record<string, unknown>;
}

export interface ManifestResponse {
  project_id: number;
  manifest: Record<string, unknown>;
}

export interface CharacterSummary {
  id: number;
  project_id: number;
  name: string;
  role: string;
  identity_summary: string | null;
  age_range: string | null;
  style_archetype: string | null;
  movement_style: string | null;
  is_locked: boolean;
  identity_json: Record<string, unknown>;
  reference_asset_urls: string[];
  consistency_rules_json: string[];
  created_at: string;
}

export interface CharacterListResponse {
  project_id: number;
  characters: CharacterSummary[];
}

export interface CharacterGenerateResponse {
  project_id: number;
  candidates: CharacterSummary[];
}

export interface CharacterLockResponse {
  character: CharacterSummary;
}

export interface ApplyCharacterToShotsResponse {
  project_id: number;
  character_id: number;
  updated_shot_count: number;
}

export type QuickRemixProfile = "english" | "nepali" | "hindi";
export type QuickCastPreset = "female" | "male" | "mixed";
export type QuickHeritageMode = "preserve" | "swap_to_english" | "swap_to_nepali" | "swap_to_hindi" | "mix";
export type YouTubePrivacyStatus = "private" | "unlisted" | "public";

export interface QuickProjectCreateInput {
  target_original_video_url: string;
  example_original_video_url: string;
  example_remix_video_url: string;
  remix_profile: QuickRemixProfile;
  cast_preset: QuickCastPreset;
  heritage_mode: QuickHeritageMode;
  auto_generate_plan: boolean;
  run_end_to_end: boolean;
  local_output_dir?: string;
  allow_youtube_upload: boolean;
  youtube_title?: string;
  youtube_description?: string;
  youtube_privacy_status: YouTubePrivacyStatus;
}

export interface QuickConversionOutput {
  project_id: number;
  output_video_path: string;
  output_dir: string;
  download_url: string;
  youtube_upload?: Record<string, unknown> | null;
}

export interface QuickProcessingStep {
  timestamp: string;
  stage: string;
  detail: string;
  progress: number;
}

export interface QuickConversionProgress {
  project_id: number;
  status: string;
  execution: string;
  progress: number;
  current_stage?: string | null;
  processing_steps: QuickProcessingStep[];
  started_at?: string | null;
  elapsed_seconds?: number | null;
  active_worker_threads?: number;
  output_video_path?: string | null;
  download_url?: string | null;
  execution_error?: string | null;
  youtube_upload?: Record<string, unknown> | null;
}

export interface DownloadVideoItem {
  project_id: number;
  video_title: string;
  remix_details: string;
  download_url: string;
  created_at: string;
}

export interface YouTubeVideoDownloadResult {
  youtube_video_url: string;
  video_title: string;
  output_file_path: string;
  output_dir: string;
  file_size_bytes: number;
  downloaded_at: string;
}

File: backend/app/schemas/project.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

CelebrityMode = Literal[
    "fictional_only",
    "celebrity_inspired_only",
    "licensed_real_celebrity_only",
]


class ORMBaseModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class ProjectCreate(ORMBaseModel):
    target_original_video_url: HttpUrl
    example_original_video_url: HttpUrl
    example_remix_video_url: HttpUrl

    character_style: str = Field(min_length=1, max_length=255)
    region_style_swap: str = Field(min_length=1, max_length=255)
    gender_mix: str = Field(min_length=1, max_length=255)
    age_group: str = Field(min_length=1, max_length=255)
    ethnic_cultural_direction: str = Field(min_length=1, max_length=255)
    # Real celebrity likeness requires legal rights/licensing. Default stays fictional_only for safety.
    celebrity_mode: CelebrityMode = "fictional_only"

    visual_theme: str = Field(min_length=1, max_length=255)
    costume_style: str = Field(min_length=1, max_length=255)
    lighting_style: str = Field(min_length=1, max_length=255)
    cinematic_mood: str = Field(min_length=1, max_length=255)
    dance_style: str = Field(min_length=1, max_length=255)
    energy_level: str = Field(min_length=1, max_length=255)
    camera_style: str = Field(min_length=1, max_length=255)

    preserve_melody: bool = True
    remix_genre: str = Field(min_length=1, max_length=255)
    beat_intensity: str = Field(min_length=1, max_length=255)
    vocal_handling: str = Field(min_length=1, max_length=255)


class ProjectSummary(ORMBaseModel):
    id: int
    target_original_video_url: str
    remix_genre: str
    celebrity_mode: CelebrityMode
    status: str
    created_at: datetime


class CharacterBible(ORMBaseModel):
    cast_name: str
    aliases: list[str]
    persona_summary: str
    styling_notes: list[str]
    movement_notes: list[str]


class CharacterProfile(ORMBaseModel):
    name: str
    role: str
    identity: dict
    references: list[str]


class Scene(ORMBaseModel):
    scene_number: int
    title: str
    setting: str
    visual_focus: str
    choreography_note: str


class ScenePrompt(ORMBaseModel):
    scene_number: int
    prompt: str


class ProjectPlanResponse(ORMBaseModel):
    transformation_summary: str
    character_bible: CharacterBible
    storyboard_scenes: list[Scene]
    scene_prompts: list[ScenePrompt]
    editing_plan: list[str]
    consistency_rules: list[str]
    manifest: dict


class ProjectListItem(ProjectSummary):
    pass


class ProjectDetail(ProjectSummary):
    example_original_video_url: str
    example_remix_video_url: str
    character_style: str
    region_style_swap: str
    gender_mix: str
    age_group: str
    ethnic_cultural_direction: str
    visual_theme: str
    costume_style: str
    lighting_style: str
    cinematic_mood: str
    dance_style: str
    energy_level: str
    camera_style: str
    preserve_melody: bool
    beat_intensity: str
    vocal_handling: str
    config_json: Optional[dict]
    transformation_summary: Optional[str]
    character_bible: Optional[CharacterBible]
    storyboard_scenes: Optional[list[Scene]]
    scene_prompts: Optional[list[ScenePrompt]]
    editing_plan: Optional[list[str]]
    consistency_rules: Optional[list[str]]
    manifest: Optional[dict]
    updated_at: datetime


class RemixProjectRead(ProjectDetail):
    pass


class ManifestResponse(ORMBaseModel):
    project_id: int
    manifest: dict
